README.md
# cnpj

Serverless application to fetch CNPJ data from [Casa dos Dados](https://casadosdados.com.br/solucao/cnpj/pesquisa-avancada).

## Usage

### `GET`

Get data without any filter.

```sh
curl https://cnpj.rafaelfranco.com
```

### `POST `

Allow filter `body` to be sent with the request. See body format below.

```sh
curl --location --request POST 'https://cnpj.rafaelfranco.com' \
--header 'Content-Type: application/json' \
--data-raw '{
  "page": 2
}'
```

## Default body

A default body is included with the request, and can be changed [here](https://github.com/rfoel/cnpj/blob/main/handlers/utils.ts#L1).

```json
{
  "query": {
    "termo": [],
    "atividade_principal": [],
    "natureza_juridica": [],
    "uf": [],
    "municipio": [],
    "situacao_cadastral": "ATIVA",
    "cep": [],
    "ddd": []
  },
  "range_query": {
    "data_abertura": {
      "lte": null,
      "gte": null
    },
    "capital_social": {
      "lte": null,
      "gte": null
    }
  },
  "extras": {
    "somente_mei": false,
    "excluir_mei": false,
    "com_email": false,
    "incluir_atividade_secundaria": false,
    "com_contato_telefonico": false,
    "somente_fixo": false,
    "somente_celular": false,
    "somente_matriz": false,
    "somente_filial": false
  },
  "page": 1
}
```

> This is the exact body that [Casa dos Dados](https://casadosdados.com.br/solucao/cnpj/pesquisa-avancada) sends in its request, so you can just tweak it for your liking.


CHANGELOG.md
## [2.1.4](https://github.com/rfoel/cnpj/compare/v2.1.3...v2.1.4) (2022-10-26)


### Bug Fixes

* bundle chrome-aws-lambda ([2f5c378](https://github.com/rfoel/cnpj/commit/2f5c3787977730c071063bb113c41bcdbad95eeb))

## [2.1.3](https://github.com/rfoel/cnpj/compare/v2.1.2...v2.1.3) (2022-05-16)


### Bug Fixes

* log response ([fae53ef](https://github.com/rfoel/cnpj/commit/fae53ef0c22ce09dc23b53b79bb52869679701d3))

## [2.1.2](https://github.com/rfoel/cnpj/compare/v2.1.1...v2.1.2) (2022-05-16)


### Bug Fixes

* log error ([8e609bb](https://github.com/rfoel/cnpj/commit/8e609bb1dff536878d82ab55a55cb4861c913052))

## [2.1.1](https://github.com/rfoel/cnpj/compare/v2.1.0...v2.1.1) (2022-05-16)


### Bug Fixes

* add logs ([e1e2930](https://github.com/rfoel/cnpj/commit/e1e2930a6ae2d579867d3d757ef5811b3fd28606))

# [2.1.0](https://github.com/rfoel/cnpj/compare/v2.0.4...v2.1.0) (2022-05-13)


### Features

* use node v16 ([c4a57cc](https://github.com/rfoel/cnpj/commit/c4a57cc0d4f1670c525c99c1e2cf32cebcc67ad0))

## [2.0.4](https://github.com/rfoel/cnpj/compare/v2.0.3...v2.0.4) (2022-05-09)


### Bug Fixes

* add get method to api ([d442a1a](https://github.com/rfoel/cnpj/commit/d442a1a2efbeabf9c4baee0fb5bfbd70ca081a88))

## [2.0.3](https://github.com/rfoel/cnpj/compare/v2.0.2...v2.0.3) (2022-05-09)


### Bug Fixes

* add stage to provider options ([875eb7c](https://github.com/rfoel/cnpj/commit/875eb7c2c93ff6fbca55ab393fe2df87dbc04ea8))

## [2.0.2](https://github.com/rfoel/cnpj/compare/v2.0.1...v2.0.2) (2022-05-09)


### Bug Fixes

* fix deploy ([736d551](https://github.com/rfoel/cnpj/commit/736d5519987fc50cb2f5ade108fbcff439335b19))

## [2.0.1](https://github.com/rfoel/cnpj/compare/v2.0.0...v2.0.1) (2022-05-04)


### Bug Fixes

* use post and return request body ([8ae2083](https://github.com/rfoel/cnpj/commit/8ae2083c664aeb7f4d3bf3a58197e9891d9a67e6))

## [2.0.1](https://github.com/rfoel/cnpj/compare/v2.0.0...v2.0.1) (2022-05-04)


### Bug Fixes

* use post and return request body ([8ae2083](https://github.com/rfoel/cnpj/commit/8ae2083c664aeb7f4d3bf3a58197e9891d9a67e6))

# [2.0.0](https://github.com/rfoel/cnpj/compare/v1.0.1...v2.0.0) (2022-05-03)


### Features

* use serverless framework ([68b2433](https://github.com/rfoel/cnpj/commit/68b243383d88a74884905272b344df8bd42fdb53))


### BREAKING CHANGES

* also changes the api, now we receive a full body from the user

## [1.0.1](https://github.com/rfoel/cnpj/compare/v1.0.0...v1.0.1) (2022-05-02)


### Bug Fixes

* add array of user agents ([14e85b0](https://github.com/rfoel/cnpj/commit/14e85b098fa44e6ef3ff7f81ddc97726bb2a4250))

# 1.0.0 (2022-03-08)


### Bug Fixes

* add https port to security group ([da9a75e](https://github.com/rfoel/cnpj/commit/da9a75ede6e7d0f4881b0c47997572830923a8d2))
* add logs ([a476aeb](https://github.com/rfoel/cnpj/commit/a476aeb0c2ac10f678fbdfd345b3162228ffed2a))
* add try catch ([bf92583](https://github.com/rfoel/cnpj/commit/bf9258312c161c22bb5f98de3a1a2657ec4e5172))
* fix cdk script ([4e3c72e](https://github.com/rfoel/cnpj/commit/4e3c72e0f5b389aa89a644d5ee2de9473ff08868))
* fix response ([d371c39](https://github.com/rfoel/cnpj/commit/d371c39b3b3f266b8602e57554017beadf254020))
* fix ts error ([c803cbc](https://github.com/rfoel/cnpj/commit/c803cbcb99e2f9341cd06ef824d490f6a1fc5c0b))
* make browser headless ([e9938c6](https://github.com/rfoel/cnpj/commit/e9938c61f35463bd4f81bae9f893d35f9d12b724))
* move page response up ([fe04357](https://github.com/rfoel/cnpj/commit/fe04357f642405e59b2318f52a8c63c9f9d4af9b))
* open browser on first request ([9e23443](https://github.com/rfoel/cnpj/commit/9e23443fea34c0d33f0cde66af76580625930198))
* pagination check ([7e35b70](https://github.com/rfoel/cnpj/commit/7e35b70ffb97ca67e16871742b99e04dabc81884))
* reload page on every request? ([c39e13e](https://github.com/rfoel/cnpj/commit/c39e13e5c6d9da8b1b9cdfa44101fe75e48c897a))
* set headless to true ([b40bbdc](https://github.com/rfoel/cnpj/commit/b40bbdcf5f9c174e9bc57c910fe963fc5b836c66))
* use ubuntu instance ([c2b4c91](https://github.com/rfoel/cnpj/commit/c2b4c91c523469df87b3a2fc9d7566423d3da4b7))


### Features

* add city filter ([d3e30c9](https://github.com/rfoel/cnpj/commit/d3e30c9f3f0d7b4fa72fe8309978b2965e198343))
* add MEI filter ([7376133](https://github.com/rfoel/cnpj/commit/7376133ccc5cb6033db29610d9e9d36f291bc4ce))
* create a tab for each request ([904834d](https://github.com/rfoel/cnpj/commit/904834ddcde61c54e964bd96f2468ae55b6c1094))
* install nginx and certbot ([8aaa945](https://github.com/rfoel/cnpj/commit/8aaa9454923a19d5c5e230b5098857f0bace1baa))
* use lambda instead of ec2 ([aeadde8](https://github.com/rfoel/cnpj/commit/aeadde8bec6276c626e36073fc37d014c4f185d2))


### BREAKING CHANGES

* use lambda instead of ec2


src/index.ts
import { PuppeteerBlocker } from '@cliqz/adblocker-puppeteer'
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda'
import chromium from 'chrome-aws-lambda'
import fetch from 'cross-fetch'
import * as puppeteer from 'puppeteer-core'

import userAgents from './userAgents'

let browserPage: puppeteer.Page

const root = 'https://casadosdados.com.br/solucao/cnpj/pesquisa-avancada/'

const launch = async () => {
  if (!browserPage || browserPage.isClosed()) {
    const browser = await chromium.puppeteer.launch({
      args: chromium.args,
      executablePath: await chromium.executablePath,
      headless: chromium.headless,
    })
    browserPage = await browser.newPage()
    await browserPage.setUserAgent(
      userAgents[Math.floor(Math.random() * userAgents.length)],
    )
    await browserPage.setViewport({
      width: 1000,
      height: 600,
    })
    const blocker = await PuppeteerBlocker.fromPrebuiltAdsAndTracking(fetch)
    await blocker.enableBlockingInPage(browserPage)
  }
}

const getCurrentPage = (): Promise<string> =>
  browserPage
    .$('.pagination-link.is-current')
    .then((element) => element?.getInnerText() as unknown as string)

const delay = (ms?: 500) => new Promise((resolve) => setTimeout(resolve, ms))

let result: APIGatewayProxyResult

const waitForResult = async (): Promise<APIGatewayProxyResult> => {
  if (!result) {
    await delay()
    return waitForResult()
  }
  await browserPage.close()
  return result
}

export const handler = async (
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> => {
  try {
    await launch()
    const { excluir_mei, municipio, page, somente_mei } =
      event.queryStringParameters || {}

    await browserPage.goto(root)

    let currentPage = '1'

    browserPage.on('response', async (response) => {
      if (
        response.request().method() === 'POST' &&
        response.url().includes('/search') &&
        (!page || (page && currentPage === page))
      ) {
        const json = await response.json()
        console.log(`Got response with ${json?.data?.count} results`)
        result = {
          statusCode: 200,
          body: JSON.stringify(json, null, 2),
        }
      }
    })

    console.log('Selecting state...')
    await browserPage.type('[placeholder="Selecione o estado"]', 'Minas Gerais')

    await browserPage
      .waitForText('MG - Minas Gerais')
      .then((element) => element.click())

    await browserPage.click('h1')
    console.log('State selected')
    console.log('Selection city...')
    await browserPage.click('[placeholder="Selecione um município"]')
    const options = await browserPage.$$(
      '.dropdown-menu.is-opened-top .dropdown-item',
    )

    if (typeof municipio === 'string') {
      await browserPage.type(
        '[placeholder="Selecione um município"]',
        municipio,
      )
      await browserPage
        .$('.dropdown-menu.is-opened-top .dropdown-item')
        .then((element) => element?.click())
    } else {
      await options[Math.floor(Math.random() * options.length)].click()
    }
    await browserPage.click('h1')
    console.log('City selected')

    if (excluir_mei === 'true') {
      console.log('Excluding MEI')
      await browserPage
        .waitForText('Excluir MEI')
        .then((element) => element.click())
    }
    if (somente_mei === 'true') {
      console.log('Only MEI')
      await browserPage
        .waitForText('Somente MEI')
        .then((element) => element.click())
    }

    console.log('Hitting search...')
    await browserPage.click('button.is-medium.is-success')
    await browserPage.waitForResponse((response) =>
      response.url().includes('/search'),
    )
    console.log('Search complete')

    currentPage = await getCurrentPage()

    if (page && currentPage !== page) {
      await browserPage.evaluate(
        () =>
          new Promise((resolve) => {
            const element = document.querySelector(
              '.pagination-link.is-current',
            )
            element?.scrollIntoView()
            resolve(1)
          }),
      )

      console.log(`Will navigate to page ${page}`)

      while (currentPage !== page) {
        const selector = `.pagination-link.pagination-${
          page && Number(page) < Number(currentPage) ? 'previous' : 'next'
        }`
        await browserPage.click(selector)
        currentPage = await getCurrentPage()
        console.log(`Navigated to page ${currentPage}`)
      }
    }

    return waitForResult()
  } catch (error) {
    console.log(error)
    return { statusCode: 500, body: JSON.stringify(error, null, 2) }
  }
}


stacks/CNPJStack.ts
import * as sst from '@serverless-stack/resources'
import { LayerVersion } from 'aws-cdk-lib/aws-lambda'

const layerArn =
  'arn:aws:lambda:us-east-1:764866452798:layer:chrome-aws-lambda:25'

export default class MyStack extends sst.Stack {
  constructor(scope: sst.App, id: string, props?: sst.StackProps) {
    super(scope, id, props)

    const prefix = scope.stage === 'prod' ? '' : `${scope.stage}.`
    const domainName = `${prefix}cnpj.rfoel.dev`

    const layer = LayerVersion.fromLayerVersionArn(this, 'Layer', layerArn)

    const api = new sst.Api(this, 'Api', {
      customDomain: {
        domainName,
        hostedZone: 'rfoel.dev',
      },
      routes: {
        'GET /': {
          function: {
            handler: 'src/index.handler',
            timeout: 30,
            layers: [layer],
            bundle: { externalModules: ['chrome-aws-lambda'] },
          },
        },
      },
    })

    this.addOutputs({
      ApiEndpoint: api.url,
    })
  }
}


lib/ec2-cdk-stack.ts
import * as cdk from 'aws-cdk-lib'
import * as ec2 from 'aws-cdk-lib/aws-ec2'
import * as iam from 'aws-cdk-lib/aws-iam'
import { KeyPair } from 'cdk-ec2-key-pair'
import { Construct } from 'constructs'

export class Ec2CdkStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    const key = new KeyPair(this, 'KeyPair', {
      name: 'cdk-keypair',
      description: 'Key Pair created with CDK Deployment',
    })

    const defaultVpc = ec2.Vpc.fromLookup(this, 'VPC', {
      isDefault: true,
    })

    const securityGroup = new ec2.SecurityGroup(this, 'SecurityGroup', {
      vpc: defaultVpc,
      allowAllOutbound: true,
    })
    securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(22))
    securityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80),
      'httpIpv4',
    )
    securityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(443),
      'httpIpv4',
    )
    securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(3000))

    const role = new iam.Role(this, 'ec2Role', {
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
    })

    role.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName(
        'AmazonSSMManagedInstanceCore',
      ),
    )

    const userData = ec2.UserData.forLinux()
    userData.addCommands(
      'apt-get update -y',
      'apt-get install -y git awscli ec2-instance-connect',
      'until git clone https://github.com/aws-quickstart/quickstart-linux-utilities.git; do echo "Retrying"; done',
      'cd /quickstart-linux-utilities',
      'source quickstart-cfn-tools.source',
      'qs_update-os || qs_err',
      'qs_bootstrap_pip || qs_err',
      'qs_aws-cfn-bootstrap || qs_err',
      'mkdir -p /opt/aws/bin',
      'ln -s /usr/local/bin/cfn-* /opt/aws/bin/',
    )

    const machineImage = ec2.MachineImage.fromSsmParameter(
      '/aws/service/canonical/ubuntu/server/focal/stable/current/amd64/hvm/ebs-gp2/ami-id',
      {
        os: ec2.OperatingSystemType.LINUX,
        userData,
      },
    )

    const ec2Instance = new ec2.Instance(this, 'Instance', {
      vpc: defaultVpc,
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T2,
        ec2.InstanceSize.MICRO,
      ),
      machineImage: machineImage,
      securityGroup: securityGroup,
      keyName: key.keyPairName,
      role: role,
      init: ec2.CloudFormationInit.fromElements(
        ec2.InitCommand.shellCommand('sudo apt update'),
        ec2.InitCommand.shellCommand('sudo apt install make nginx -y'),
        ec2.InitCommand.shellCommand(
          'sudo snap install core; sudo snap refresh core',
        ),
        ec2.InitCommand.shellCommand('sudo snap install --classic certbot'),
        ec2.InitCommand.shellCommand(
          'sudo ln -s /snap/bin/certbot /usr/bin/certbot',
        ),
        ec2.InitCommand.shellCommand(
          'curl -L https://git.io/n-install | bash -s -- -y',
        ),
        ec2.InitCommand.shellCommand('. ~/.bashrc'),
        ec2.InitCommand.shellCommand('npm i -g pm2'),
        ec2.InitCommand.shellCommand(
          'yes | npx playwright install-deps chromium',
        ),
      ),
    })

    new cdk.CfnOutput(this, 'IP Address', {
      value: ec2Instance.instancePublicIp,
    })
    new cdk.CfnOutput(this, 'Key Name', { value: key.keyPairName })
    new cdk.CfnOutput(this, 'Download Key Command', {
      value:
        'aws secretsmanager get-secret-value --secret-id ec2-ssh-key/cdk-keypair/private --query SecretString --output text > cdk-key.pem && chmod 400 cdk-key.pem',
    })
    new cdk.CfnOutput(this, 'ssh command', {
      value:
        'ssh -i cdk-key.pem -o IdentitiesOnly=yes ubuntu@' +
        ec2Instance.instancePublicIp,
    })
  }
}


handlers/cnpj.ts
import type { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda'
import chromium from 'chrome-aws-lambda'
import type { Browser, Page } from 'puppeteer-core'

import { defaultBody, userAgents } from './utils'

let browser: Browser
let page: Page

const launch = async () => {
  if (!page || page.isClosed()) {
    if (!browser) {
      browser = await chromium.puppeteer.launch({
        args: chromium.args,
        executablePath: await chromium.executablePath,
        headless: chromium.headless,
      })
    }
    page = await browser.newPage()
    await page.setUserAgent(
      userAgents[Math.floor(Math.random() * userAgents.length)],
    )
  }
}

const close = async () => {
  if (page && !page.isClosed()) {
    await page.close()
  }
}

export const handler = async (
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> => {
  console.log('REQUEST', JSON.stringify(event, null, 2))

  let data
  let response

  try {
    await launch()

    await page.goto(
      'https://casadosdados.com.br/solucao/cnpj/pesquisa-avancada',
    )

    await page.evaluate((body) => {
      // @ts-expect-error for some reason, typescript won't pick up the fetch function from puppeteer
      void fetch('https://api.casadosdados.com.br/v2/public/cnpj/search', {
        headers: {
          accept: 'application/json, text/plain, */*',
          'accept-language': 'pt-BR,pt;q=0.9,en;q=0.8,de;q=0.7',
          'content-type': 'application/json;charset=UTF-8',
          'sec-ch-ua':
            '" Not A;Brand";v="99", "Chromium";v="100", "Google Chrome";v="100"',
          'sec-ch-ua-mobile': '?0',
          'sec-ch-ua-platform': '"macOS"',
          'sec-fetch-dest': 'empty',
          'sec-fetch-mode': 'cors',
          'sec-fetch-site': 'same-site',
        },
        referrer: 'https://casadosdados.com.br/',
        referrerPolicy: 'strict-origin-when-cross-origin',
        body: typeof body === 'object' ? JSON.stringify(body) : body,
        method: 'POST',
        mode: 'cors',
        credentials: 'omit',
      })
    }, event?.body || defaultBody)

    data = await page.waitForResponse(
      (response) =>
        response.url().includes('/cnpj/search') && response.status() === 200,
    )
    const parsedData = await data.json()

    await close()

    response = {
      statusCode: 200,
      body: JSON.stringify(parsedData),
    }
  } catch (error) {
    console.log('ERROR', error)
    await close()
    response = { statusCode: 500, body: 'See cloudwatch for the error' }
  }

  console.log('RESPONSE', JSON.stringify(response, null, 2))

  return response
}
